File: bootstrap_sampling.R
# Bootstrap Sampling in R

# Function to perform bootstrap sampling
bootstrap_sampling <- function(data, n_bootstrap) {
  sample_means <- numeric(n_bootstrap)
  for (i in 1:n_bootstrap) {
    sample <- sample(data, length(data), replace = TRUE)
    sample_means[i] <- median(sample)
  }
  return(sample_means)
}

# Function to calculate standard error
standard_error <- function(sample_means) {
  return(sd(sample_means) / sqrt(length(sample_means)))
}

# Generating distributions
gen_start_time <- Sys.time()
set.seed(123)
positively_skewed <- rexp(1000, rate = 0.1)
symmetric <- rnorm(1000, mean = 50, sd = 10)
negatively_skewed <- sort(rexp(1000, rate = 0.1), decreasing = TRUE)
gen_end_time <- Sys.time()
gen_time <- gen_end_time - gen_start_time

# Bootstrap sampling
n_bootstrap <- 10000
pos_boot_start_time <- Sys.time()
bootstrap_positively_skewed <- bootstrap_sampling(positively_skewed, n_bootstrap)
pos_boot_end_time <- Sys.time()
pos_bootstrap_time <- pos_boot_end_time - pos_boot_start_time

sym_boot_start_time <- Sys.time()
bootstrap_symmetric <- bootstrap_sampling(symmetric, n_bootstrap)
sym_boot_end_time <- Sys.time()
sym_bootstrap_time <- sym_boot_end_time - sym_boot_start_time

neg_boot_start_time <- Sys.time()
bootstrap_negatively_skewed <- bootstrap_sampling(negatively_skewed, n_bootstrap)
neg_boot_end_time <- Sys.time()
neg_bootstrap_time <- neg_boot_end_time - neg_boot_start_time

# Calculating and printing standard error of the median
se_calc_start_time <- Sys.time()
se_positively_skewed <- standard_error(bootstrap_positively_skewed)
se_symmetric <- standard_error(bootstrap_symmetric)
se_negatively_skewed <- standard_error(bootstrap_negatively_skewed)
se_calc_end_time <- Sys.time()
se_calc_time <- se_calc_end_time - se_calc_start_time

cat("Standard Error of the Median for n_bootstrap =", n_bootstrap, ":\n")
cat("Positively Skewed: ", se_positively_skewed, "\n")
cat("Symmetric: ", se_symmetric, "\n")
cat("Negatively Skewed: ", se_negatively_skewed, "\n")

cat("Runtimes:\n")
cat("Generation time: ", gen_time, " seconds\n")
cat("Positively Skewed Bootstrap sampling time: ", pos_bootstrap_time, " seconds\n")
cat("Symmetric Bootstrap sampling time: ", sym_bootstrap_time, " seconds\n")
cat("Negatively Skewed Bootstrap sampling time: ", neg_bootstrap_time, " seconds\n")
cat("Calculation time: ", se_calc_time, " seconds\n")

File: bootstrap_sampling.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// Function to perform bootstrap sampling
func bootstrapSampling(data []float64, nBootstrap int) []float64 {
	sampleMeans := make([]float64, nBootstrap)
	var wg sync.WaitGroup
	for i := 0; i < nBootstrap; i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			sample := make([]float64, len(data))
			for j := range sample {
				sample[j] = data[rand.Intn(len(data))]
			}
			sampleMeans[i] = median(sample)
		}(i)
	}
	wg.Wait()
	return sampleMeans
}

// Function to calculate median
func median(data []float64) float64 {
	middle := len(data) / 2
	if len(data)%2 == 0 {
		return (data[middle-1] + data[middle]) / 2
	}
	return data[middle]
}

// Function to calculate standard error
func standardError(sampleMeans []float64) float64 {
	var sum, mean float64
	for _, value := range sampleMeans {
		sum += value
	}
	mean = sum / float64(len(sampleMeans))
	var varianceSum float64
	for _, value := range sampleMeans {
		varianceSum += (value - mean) * (value - mean)
	}
	variance := varianceSum / float64(len(sampleMeans)-1)
	return sqrt(variance / float64(len(sampleMeans)))
}

// Function to calculate square root
func sqrt(value float64) float64 {
	return math.Sqrt(value)
}

func main() {
	rand.Seed(time.Now().UnixNano())
	data := generateDistributions()
	nBootstrap := 10000

	for _, dist := range data {
		bootstrapResults := bootstrapSampling(dist.values, nBootstrap)
		se := standardError(bootstrapResults)
		fmt.Printf("Distribution: %s, Standard Error of the Median: %f\n", dist.name, se)
	}
}

// Distribution struct
type Distribution struct {
	name   string
	values []float64
}

// Function to generate distributions
func generateDistributions() []Distribution {
	return []Distribution{
		{"Positively Skewed", generatePositivelySkewed(1000)},
		{"Symmetric", generateSymmetric(1000)},
		{"Negatively Skewed", generateNegativelySkewed(1000)},
	}
}

// Function to generate positively skewed distribution
func generatePositivelySkewed(size int) []float64 {
	dist := make([]float64, size)
	for i := range dist {
		dist[i] = rand.ExpFloat64()
	}
	return dist
}

// Function to generate symmetric distribution
func generateSymmetric(size int) []float64 {
	dist := make([]float64, size)
	for i := range dist {
		dist[i] = rand.NormFloat64() * 10 + 50
	}
	return dist
}

// Function to generate negatively skewed distribution
func generateNegativelySkewed(size int) []float64 {
	dist := generatePositivelySkewed(size)
	sort.Sort(sort.Reverse(sort.Float64Slice(dist)))
	return dist
}
